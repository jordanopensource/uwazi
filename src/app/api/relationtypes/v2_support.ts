import { ObjectId } from 'mongodb';
import { DefaultSettingsDataSource } from 'api/settings.v2/database/data_source_defaults';
import { DefaultRelationshipDataSource } from 'api/relationships.v2/database/data_source_defaults';
import { CreateTemplateService } from 'api/templates.v2/services/service_factories';
import { DefaultTransactionManager } from 'api/common.v2/database/data_source_defaults';

const getNewRelationshipCount = async (id: ObjectId) => {
  const transactionManager = DefaultTransactionManager();
  const newRelationshipsAllowed =
    await DefaultSettingsDataSource(transactionManager).readNewRelationshipsAllowed();
  const relationshipsDataSource = DefaultRelationshipDataSource(transactionManager);

  return newRelationshipsAllowed ? relationshipsDataSource.countByType(id.toString()) : 0;
};

const relationTypeIsUsedInQueries = async (id: ObjectId): Promise<boolean> => {
  const transactionManager = DefaultTransactionManager();
  const newRelationshipsAllowed =
    await DefaultSettingsDataSource(transactionManager).readNewRelationshipsAllowed();
  if (!newRelationshipsAllowed) return false;

  const createTemplateService = await CreateTemplateService();
  const isUsed = await createTemplateService.relationTypeIsUsedInQueries(id.toString());

  return isUsed;
};

export { getNewRelationshipCount, relationTypeIsUsedInQueries };
