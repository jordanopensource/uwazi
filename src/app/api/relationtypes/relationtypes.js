import translations from 'api/i18n/translations';
import relationships from 'api/relationships/relationships';
import templates from 'api/templates';
import { createError } from 'api/utils';
import { ContextType } from 'shared/translationSchema';
import { generateNames, getUpdatedNames, getDeletedProperties } from '../templates/utils';
import model from './model';
import * as v2 from './v2_support';

const checkDuplicated = relationtype =>
  model.get().then(response => {
    const duplicated = response.find(entry => {
      const sameEntity = entry._id.equals(relationtype._id);
      const sameName = entry.name.trim().toLowerCase() === relationtype.name.trim().toLowerCase();
      return sameName && !sameEntity;
    });

    if (duplicated) {
      return Promise.reject('duplicated_entry');
    }
  });

function _save(relationtype) {
  const values = {};
  values[relationtype.name] = relationtype.name;
  relationtype.properties.forEach(property => {
    values[property.label] = property.label;
  });
  return model
    .save(relationtype)
    .then(response =>
      translations
        .addContext(response._id, relationtype.name, values, ContextType.relationshipType)
        .then(() => response)
    );
}

const updateTranslation = (currentTemplate, template) => {
  const currentProperties = currentTemplate.properties;
  const newProperties = template.properties;

  const updatedLabels = getUpdatedNames(
    {
      prop: 'label',
      filterBy: '_id',
    },
    currentProperties,
    newProperties
  ).update;
  if (currentTemplate.name !== template.name) {
    updatedLabels[currentTemplate.name] = template.name;
  }
  const deletedPropertiesByLabel = getDeletedProperties(
    currentProperties,
    newProperties,
    '_id',
    'label'
  );
  const context = template.properties.reduce((ctx, prop) => {
    ctx[prop.label] = prop.label;
    return ctx;
  }, {});

  context[template.name] = template.name;

  return translations.updateContext(
    { id: currentTemplate._id.toString(), label: template.name, type: 'Connection' },
    updatedLabels,
    deletedPropertiesByLabel,
    context
  );
};

function _update(newTemplate) {
  return model.getById({ _id: newTemplate._id }).then(currentTemplate => {
    updateTranslation(currentTemplate, newTemplate);
    relationships.updateMetadataProperties(newTemplate, currentTemplate);
    return model.save(newTemplate);
  });
}

async function validateTypeInTemplates(id) {
  const templatesUsingType = await templates.findUsingRelationTypeInProp(id);
  if (templatesUsingType.length > 0) {
    const templatesNames = templatesUsingType.map(t => t.name).join(', ');
    throw createError(`Cannot delete type being used in templates: ${templatesNames}`, 400);
  }
}

export default {
  get(query) {
    return model.get(query);
  },

  getById(id) {
    return model.getById(id);
  },

  async save(relationtype) {
    relationtype.properties = await generateNames(relationtype.properties || []);

    return checkDuplicated(relationtype).then(() => {
      if (!relationtype._id) {
        return _save(relationtype);
      }
      return _update(relationtype);
    });
  },

  async delete(id) {
    await validateTypeInTemplates(id);

    const connectionCount = await relationships.countByRelationType(id);
    const newRelationshipCount = await v2.getNewRelationshipCount(id);
    const relationTypeIsUsedInQueries = await v2.relationTypeIsUsedInQueries(id);

    if (connectionCount === 0 && newRelationshipCount === 0 && !relationTypeIsUsedInQueries) {
      await translations.deleteContext(id);
      await model.delete(id);
      return true;
    }

    return false;
  },
};
