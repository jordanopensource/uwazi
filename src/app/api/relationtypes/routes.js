import relationtypes from 'api/relationtypes/relationtypes';
import { ObjectIdAsString } from 'api/utils/ajvSchemas';
import { validation } from '../utils';
import needsAuthorization from '../auth/authMiddleware';

export default app => {
  app.post(
    '/api/relationtypes',
    needsAuthorization(),
    validation.validateRequest({
      type: 'object',
      properties: {
        body: {
          type: 'object',
          properties: {
            _id: ObjectIdAsString,
            __v: { type: 'number' },
            name: { type: 'string' },
            properties: {
              type: 'array',
              items: {
                type: 'object',
                properties: {
                  _id: { type: 'string' },
                  __v: { type: 'number' },
                  localID: { type: 'string' },
                  id: { type: 'string' },
                  label: { type: 'string' },
                  type: { type: 'string' },
                  content: { type: 'string' },
                  name: { type: 'string' },
                  filter: { type: 'boolean' },
                  sortable: { type: 'boolean' },
                  showInCard: { type: 'boolean' },
                  prioritySorting: { type: 'boolean' },
                  nestedProperties: { type: 'array' },
                },
              },
            },
          },
        },
      },
      required: ['body'],
    }),
    (req, res, next) => {
      relationtypes
        .save(req.body)
        .then(response => res.json(response))
        .catch(next);
    }
  );

  app.get(
    '/api/relationtypes',
    validation.validateRequest({
      type: 'object',
      properties: {
        query: {
          type: 'object',
          properties: {
            _id: ObjectIdAsString,
          },
        },
      },
      required: ['query'],
    }),
    (req, res, next) => {
      if (req.query._id) {
        return relationtypes
          .getById(req.query._id)
          .then(response => res.json({ rows: [response] }))
          .catch(next);
      }
      relationtypes
        .get()
        .then(response => res.json({ rows: response }))
        .catch(next);
    }
  );

  app.delete(
    '/api/relationtypes',
    needsAuthorization(),
    validation.validateRequest({
      type: 'object',
      properties: {
        query: {
          type: 'object',
          properties: {
            _id: ObjectIdAsString,
          },
          required: ['_id'],
        },
      },
      required: ['query'],
    }),
    (req, res, next) => {
      relationtypes
        .delete(req.query._id)
        .then(response => res.json(response))
        .catch(next);
    }
  );
};
