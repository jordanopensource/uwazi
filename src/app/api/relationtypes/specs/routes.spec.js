import relationtypesRoutes from 'api/relationtypes/routes.js';
import instrumentRoutes from 'api/utils/instrumentRoutes';
import relationtypes from 'api/relationtypes/relationtypes';

describe('relationtypes routes', () => {
  let routes;
  const mockRequest = new Promise(resolve => {
    resolve({ relationtypes: 'response' });
  });

  beforeEach(() => {
    routes = instrumentRoutes(relationtypesRoutes);
  });

  describe('GET', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/relationtypes')).toMatchSnapshot();
    });

    it('should ask relationtypes for all documents', async () => {
      jest.spyOn(relationtypes, 'get').mockReturnValue(mockRequest);
      const response = await routes.get('/api/relationtypes', { query: {} });
      expect(relationtypes.get).toHaveBeenCalled();
      expect(response).toEqual({ rows: { relationtypes: 'response' } });
    });

    describe('when passing id', () => {
      it('should ask for the specific relationtype', async () => {
        jest.spyOn(relationtypes, 'getById').mockReturnValue(mockRequest);
        const response = await routes.get('/api/relationtypes', { query: { _id: 'someId' } });
        expect(relationtypes.getById).toHaveBeenCalledWith('someId');
        expect(response).toEqual({ rows: [{ relationtypes: 'response' }] });
      });
    });
  });

  describe('DELETE', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/relationtypes')).toMatchSnapshot();
    });

    it('should delete the relationtype', async () => {
      jest.spyOn(relationtypes, 'delete').mockReturnValue(mockRequest);
      const response = await routes.delete('/api/relationtypes', {
        query: { _id: 'someId', _rev: 'latest' },
      });
      expect(relationtypes.delete).toHaveBeenCalledWith('someId');
      expect(response).toEqual({ relationtypes: 'response' });
    });
  });

  describe('POST', () => {
    it('should have a validation schema', () => {
      expect(routes.post.validation('/api/relationtypes')).toMatchSnapshot();
    });

    it('should save the relationtype', async () => {
      jest.spyOn(relationtypes, 'save').mockReturnValue(mockRequest);
      const response = await routes.post('/api/relationtypes', {
        body: { name: 'my new template' },
      });
      expect(relationtypes.save).toHaveBeenCalledWith({ name: 'my new template' });
      expect(response).toEqual({ relationtypes: 'response' });
    });
  });
});
