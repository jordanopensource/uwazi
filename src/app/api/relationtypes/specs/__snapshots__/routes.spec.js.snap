// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relationtypes routes DELETE should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "_id": Object {
          "pattern": "^[0-9a-fA-F]{24}$",
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`relationtypes routes GET should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "_id": Object {
          "pattern": "^[0-9a-fA-F]{24}$",
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`relationtypes routes POST should have a validation schema 1`] = `
Object {
  "properties": Object {
    "body": Object {
      "properties": Object {
        "__v": Object {
          "type": "number",
        },
        "_id": Object {
          "pattern": "^[0-9a-fA-F]{24}$",
          "type": "string",
        },
        "name": Object {
          "type": "string",
        },
        "properties": Object {
          "items": Object {
            "properties": Object {
              "__v": Object {
                "type": "number",
              },
              "_id": Object {
                "type": "string",
              },
              "content": Object {
                "type": "string",
              },
              "filter": Object {
                "type": "boolean",
              },
              "id": Object {
                "type": "string",
              },
              "label": Object {
                "type": "string",
              },
              "localID": Object {
                "type": "string",
              },
              "name": Object {
                "type": "string",
              },
              "nestedProperties": Object {
                "type": "array",
              },
              "prioritySorting": Object {
                "type": "boolean",
              },
              "showInCard": Object {
                "type": "boolean",
              },
              "sortable": Object {
                "type": "boolean",
              },
              "type": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
    },
  },
  "required": Array [
    "body",
  ],
  "type": "object",
}
`;
